actors/ppo.py

```diff
@@ -0,0 +1,720 @@
+        self.value_scheduler = torch.optim.lr_scheduler.ExponentialLR(self.value_optimizer, gamma=0.99)
+
+        # version user data
+        self.iter_idx = 0
+        self.total_samples = 0
+
+    @property
+    def state_norm(self) -> bool:
+        return self._state_norm
+
+    @state_norm.setter
+    def state_norm(self, value: bool) -> None:
+        self._state_norm = value
+        if self._state_norm:
+            self.state_normalization = Normalization(dtype=self._dtype, nums=self._num_input)
+
+    def get_model_user_data(self) -> dict:
+        """Get user model"""
+        return {
+            "environment_implementation": self._environment_implementation,
+            "num_input": self._num_input,
+            "num_output": self._num_output,
+            "policy_network_hidden_nodes": self._policy_network_hidden_nodes,
+            "value_network_hidden_nodes": self._value_network_hidden_nodes,
+        }
+
+    def save(self, model_data_f: str) -> dict:
+        """Save the model"""
+        torch.save((self.policy_network.state_dict(), self.value_network.state_dict()), model_data_f)
+        return {"iter_idx": self.iter_idx, "total_samples": self.total_samples}
+
+    @classmethod
+    def load(
+        cls, model_id: int, version_number: int, model_user_data: dict, version_user_data: dict, model_data_f: str
+    ) -> Model:
+        """Load the model"""
+        model = PPOModel(
+            model_id=model_id,
+            version_number=version_number,
+            environment_implementation=model_user_data["environment_implementation"],
+            num_input=int(model_user_data["num_input"]),
+            num_output=int(model_user_data["num_output"]),
+            policy_network_hidden_nodes=int(model_user_data["policy_network_hidden_nodes"]),
+            value_network_hidden_nodes=int(model_user_data["value_network_hidden_nodes"]),
+        )
+
+        # Load the model parameters
+        (policy_network_state_dict, value_network_state_dict) = torch.load(model_data_f)
+        model.policy_network.load_state_dict(policy_network_state_dict)
+        model.value_network.load_state_dict(value_network_state_dict)
+
+        # Load version data
+        model.iter_idx = version_user_data["iter_idx"]
+        model.total_samples = version_user_data["total_samples"]
+        return model
+
+
+class PPOActor:
+    """PPO actor"""
+
+    def __init__(self, _cfg):
+        self._dtype = torch.float
+
+    def get_actor_classes(self):
+        """Get actor"""
+        return [PLAYER_ACTOR_CLASS]
+
+    async def impl(self, actor_session):
+        # Start a session
+        actor_session.start()
+        config = actor_session.config
+        assert config.environment_specs.num_players == 1
+        assert len(config.environment_specs.action_space.properties) == 1
+        assert config.environment_specs.action_space.properties[0].WhichOneof("type") == "box"
```

Comment:
Ideally PPO should also work for discrete and multi-discrete actions

Reply:
Agree. Upcoming version will address this issue